File: cs/src/ThePlayers.cs

using System;

namespace TddExRefactor2
{
    public class ThePlayers {
        private int[] _keys;
        private string[] _names;
        private int _count;
        private bool _readOnly;

        public ThePlayers(bool makeReadOnly) {
            this._keys = new int[0];
            this._names = new string[0];
            this._count = 0;
            this._readOnly = makeReadOnly;
        }

        public  void Add(int key, string name) 
        {
            if (!_readOnly) 
            {
                for (int i = 0; i < _count; i++)
                    if (_keys[i].Equals(key)) 
                    {
                    _names[i] = name;
                    return;
                    }
                int newSize = _count + 1;
                if (newSize > _keys.Length) 
                {
                    int[] newKeys = new int[_keys.Length + 1];
                    string[] newNames = new string[_keys.Length + 1];
                    Array.Copy(_keys, newKeys, _count);
                    Array.Copy(_names, newNames, _count);
                    _keys = newKeys;
                    _names = newNames;
                }
                _keys[_count] = key;
                _names[_count] = name;
                _count++;
            }
        }

        public  bool Remove(int key) 
        {
            if (_readOnly)
                return false;
            for (int i = 0; i < _count; i++)
                if (_keys[i].Equals(key)) 
                {
                    _keys[i] = -1;
                    _names[i] = "";
                    _count--;
                    return true;
                }
            return false;
        }

        public string get(int key) {
            var index = Array.IndexOf(_keys, key);
            if(index < 0) {
                throw new ArgumentOutOfRangeException($"Key {key} not found.");
            }
            return this._names[index];
        }

        public int count() {
            return this._count;
        }

        public int size() {
            return this._keys.Length;
        }
    }
}

File: py/src/ThePlayers.py
from .Array import Array

class ThePlayers():
  GROWTH = 5

  def __init__(self, makeReadOnly):
    self.__keys = Array(0)
    self.__names = Array(0)
    self.__count = 0
    self.__readOnly: bool = makeReadOnly

  def get(self, key:int):
    index = self.__keys.index_of(key)
    if index < 0:
      raise Exception(f'Key ${key} not found')
    return self.__names.g(index)

  def add(self, key: int, name: str):
    if not self.__readOnly:
      for i in range(self.__count):
        if self.__keys.g(i) == key:
          self.__names.s(i, name)
          return
      newsize = self.__count + 1
      if newsize > self.size():
        new_keys = Array(self.size() + self.GROWTH)
        new_names = Array(self.size() + self.GROWTH)
        Array.copy(self.__keys, new_keys, self.__count)
        Array.copy(self.__names, new_names, self.__count)
        self.__keys = new_keys
        self.__names = new_names
      self.__keys.s(self.__count, key)
      self.__names.s(self.__count, name)
      self.__count = self.__count + 1

  def remove(self, key: int):
    if self.__readOnly: 
      return False
    for i in range(self.__count):
      if self.__keys.g(i) == key:
        self.__keys.s(i, -1)
        self.__names.s(i, '')
        self.__count = self.__count - 1
        return True
    return False

  def count(self):
    return self.__count

  def size(self):
    return self.__keys.length()